SKIPJob v1beta1

EvenSandvik opened the pull request:
> ## Skiperator SKIPJob v1alpha -> v1beta1 upgrade
> ## Why?
> The reason for upgrading the api version for skipjob, is to remove the `cont<ainer` field in `spec` which is inconsistent with the `spec` in Application.
> 
> ## How?
> * We have introduced a new skiperator api version **v1beta1** where container is removed, and its fields are
>   put directly into `spec`.
> * In order to still support **v1alpha1** resource we have also created a **conversion webhook**
> * Introduced **common-types** that can be used for both v1alpha1 and v1beta1.
> * Added config for certificates for TLS communication between the Kube API server and Webhook Server in skiperator.
> 
> ## Test
> ```shell
> # start with clean slate
> kind delete cluster --name skiperator
> 
> # set up cluster
> make setup-local
> 
> # run skiperator with webhook and cert config
> make run-local
> 
> # make a namespace
> kubectl create namespace devex
> 
> # apply a v1alpha1 manifest
> kubectl apply -f samples/v1alpha1/test-job.yaml
> 
> # apply a v1beta1 manifest
> kubectl apply -f samples/v1beta1/test-job.yaml
> 
> # get the stored objects and inspect the spec of both, the two objects should be identical
> kubectl get skipjobs -n devex -o json | jq '.items[].spec'
> 
> # get the v1alpha1 manifest, should be stored as a v1beta1 resource
> kubectl get skipjobs -n devex sample-job-old -o yaml
> 
> 
> # inspect the last-applied-configuration in the metadata (shows the original v1alpha1 manifest we applied)
> kubectl get skipjobs -n devex sample-job-old -o json \
>     | jq -r '.metadata.annotations["kubectl.kubernetes.io/last-applied-configuration"]' \
>     | jq
> ```
> 
> ## Changes
> As you might have noticed, there are a lot of changes in this PR, however most of these changes are due to refactoring of import statements. These are mainly in `internal/` and `pkg/` and are described further below.
> 
> ### New types in `api/`
> * **common:** created common types based on v1alpha1 types
> * **v1alpha1:** refactor to use the common types and create a conversion function
> * **v1beta1:** define v2 api types
> 
> ### Changes in main `cmd/`
> * **skiperator** created the webhook server and additional config for webhook and certs
> 
> ### New configuration in `config/`
> * **crd**: added patch with kustomize to inject the ca-bundle and webhook config into the SKIPJob CRD
> * **cert-manager**: create Certificates and Issuer for cert manager to create certificates use by the kubeapi and webhook server
> * **webhook**: defined a Service for the webhook
> 
> ### local dev scripts in `hack/`
> * script for getting webhook certs to local file
> * script for creating Endpoint resource pointing to local webhook server
> 
> ### refactors in `internal/`
> * replace references to **v1alpha1** with **v1beta1** or **common** (common types)
> * remove `.container` references, `foo.container.bar` becomes `foo.bar`
> * define webhook setup for skipjob
> 
> ### refactors in `pkg/`
> * replace references to **v1alpha1** with **v1beta1** or **common** (common types)
> * remove `.container` references, `foo.container.bar` becomes `foo.bar`
> 
> ### testfiles in `samples/` (might remove?)
> * testfiles for skipjob with versions: **v1alpha1** and **v1beta1**
> 
> ## Summary by CodeRabbit
> * **New Features**
>   
>   * v1beta1 SKIPJob API with conversion support and webhook-based conversion (cert-manager integration)
>   * Local webhook developer workflow and helper scripts; Makefile flags for local webhook certs and host
>   * Prometheus scrape configuration added; samples for v1alpha1/v1beta1 SKIPJob
> * **Refactor**
>   
>   * Shared types consolidated for cross-version reuse; API type aliases added
>   * Installer/run improvements: explicit image loading for Istio/cert-manager and enhanced local cluster install flow
>   * Configuration moved to richer runtime config (feature flags → config), including cluster CIDR exclusions and registry secret refs
> * **Documentation**
>   
>   * Added sample manifests and cluster-exclusion examples
> 
> ✏️ Tip: You can customize this high-level summary in your review settings.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * SKIPJob v1beta1 released with richer spec, defaults, status, conversion and a mutating webhook for defaulting.

* **Improvements**
  * Local webhook tooling and Makefile targets for cert extraction, local endpoint, and test/benchmark integration.
  * Added CRD/webhook and cert-manager manifests, samples, Istio/cert-manager image handling, and TLS-enabled webhook support; bumped KIND and Kubernetes versions; configurable enableWebhooks.

* **Breaking Changes**
  * Public API reorganized around shared/common types and v1beta1 aliases — clients should migrate to v1beta1.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

evenh commented on api/v1beta1/skipjob_types.go:
```diff
@@ -0,0 +1,232 @@
+			},
+		}
+	}
+}
```
Porting existing behaviour, ignore for now

evenh commented on cmd/skiperator/main.go:
```diff
@@ -56,6 +79,16 @@ func main() {
 	encCfg.EncodeTime = zapcore.ISO8601TimeEncoder
 
 	// Set a temporary logger for the config loading, the real logger is initialized later with values from config
```
This comment should be above `ctrl.SetLogger(…)`

evenh commented on cmd/skiperator/main.go:
```diff
@@ -112,7 +145,83 @@ func main() {
 	if activeConfig.EnableProfiling {
 		pprofBindAddr = ":8281"
 	}
+	var metricsAddr string
```
Move variables to `var ()` block at the top under `Version` and `Commit`

evenh commented on cmd/skiperator/main.go:
```diff
@@ -112,7 +145,83 @@ func main() {
+			"webhook-cert-dir", *webhookCertDir, "webhook-cert-name", *webhookCertName, "webhook-key-name", *webhookKeyName)
+
+		var err error
+		webhookCertWatcher, err = certwatcher.New(certPath, keyPath)
```
```suggestion
		webhookCertWatcher, err := certwatcher.New(certPath, keyPath)
```

evenh commented on cmd/skiperator/main.go:
```diff
@@ -144,6 +254,35 @@ func main() {
+		}
+	}
+
+	if webhookCertWatcher != nil {
```
This condition is always false

evenh commented on internal/webhook/skipjob_webhook.go:
```diff
@@ -0,0 +1,69 @@
+			DefaultBackoffLimit:            v1beta1.DefaultBackoffLimit,
+			DefaultSuspend:                 v1beta1.DefaultSuspend,
+		}).
+		// TODO: Add a validator when this works...
```
Dangling TODO?

hauks1 replied:
Link to kubebuilder convention: https://github.com/kubernetes-sigs/kubebuilder/blob/1726e5882564a37ecd5d96409138e4df20b794e2/docs/book/src/cronjob-tutorial/testdata/project/internal/webhook/v1/cronjob_webhook.go#L52
They have a Defaulting and Validation webhook

evenh commented on Makefile:
```diff
@@ -60,10 +66,25 @@ kind-cluster: check-kind
+	  docker pull --platform linux/$(ARCH) "$$image"; \
+	  echo "Loading $$image into kind cluster 'skiperator'"; \
+	  docker save "$$image" | docker exec -i skiperator-control-plane ctr --namespace=k8s.io images import -; \
+	done
```
Implement this like done for cert-manager

evenh reviewed (changes requested):
Looks good overall. See comments :-)

evenh committed:
SKIPJob: introduce v1beta1 API (foundation for ArgoKit v2)

This change introduces a new API version (v1beta1) that is
functionally identical to the existing version (v1alpha1). This is
done in order to align the API shape with the Application API.

This change will make supporting SKIPJob's in ArgoKit v2 simpler and
more aligned with existing Application behaviour.

Validation and conversion logic is added to make the transition safe.

Co-authored-by: Andreas Espelund <andreas.espelund@bekk.no>
Co-authored-by: Even Sandvik <even.sandvik@bekk.no>
Co-authored-by: Hauk Storjord <hauk.storjord@bekk.no>
Co-authored-by: Thomas Berg <thomas.berg@kartverket.no>
Co-authored-by: Martin Morisbak <martin.morisbak@kartverket.no>

evenh committed:
SKIPJob: introduce v1beta1 API (foundation for ArgoKit v2)

This change introduces a new API version (v1beta1) that is
functionally identical to the existing version (v1alpha1). This is
done in order to align the API shape with the Application API.

This change will make supporting SKIPJob's in ArgoKit v2 simpler and
more aligned with existing Application behaviour.

Validation and conversion logic is added to make the transition safe.

Co-authored-by: Andreas Espelund <andreas.espelund@bekk.no>
Co-authored-by: Even Sandvik <even.sandvik@bekk.no>
Co-authored-by: Hauk Storjord <hauk.storjord@bekk.no>

Morisman96 commented on Makefile:
```diff
@@ -88,9 +145,19 @@ install-digdirator-crds:
 	@kubectl create namespace skiperator-system --context $(SKIPERATOR_CONTEXT) || true
-	@kubectl apply -f config/ --recursive --context $(SKIPERATOR_CONTEXT)
+	@kubectl apply -f config/cert-manager --context $(SKIPERATOR_CONTEXT) || true
+	@kustomize build config/crd | kubectl apply -f - --context $(SKIPERATOR_CONTEXT) || true
```
Is it not possible to just use kustomize with kubectl, less dependencies
```suggestion
	@kubectl kustomize config/crd | kubectl apply -f - --context $(SKIPERATOR_CONTEXT) || true
```

Morisman96 commented on Makefile:
```diff
@@ -22,9 +22,29 @@ KUBERNETES_VERSION          = 1.34.3
+patch-skipjob-crd:
+	@tmp_file=$$(mktemp -t skipjob-crd.XXXXXXX); \
+	normalized_file=$$(mktemp -t skipjob-crd-normalized.XXXXXXX); \
+	kustomize build --load-restrictor LoadRestrictionsNone config/crd/skipjob > "$$tmp_file"; \
```
Is it not possible to just use kustomize with kubectl, less dependencies
```suggestion
	kubectl kustomize --load-restrictor LoadRestrictionsNone config/crd/skipjob > "$$tmp_file"; \
```

Morisman96 commented on Makefile:
```diff
@@ -88,9 +145,19 @@ install-digdirator-crds:
 install-skiperator: generate
 	@kubectl create namespace skiperator-system --context $(SKIPERATOR_CONTEXT) || true
-	@kubectl apply -f config/ --recursive --context $(SKIPERATOR_CONTEXT)
+	@kubectl apply -f config/cert-manager --context $(SKIPERATOR_CONTEXT) || true
```
have encountered problems when running skiperator locally where cert-manager is not ready, leading to internal error when applying cert-manager configs. Putting a sleep seems to fix this
```suggestion
    @echo "Waiting for cert-manager to be ready before installing skiperator..."
    @sleep 10
	@kubectl apply -f config/cert-manager --context $(SKIPERATOR_CONTEXT) || true
```

Morisman96 reviewed (changes requested):
some suggestion in the make file

evenh committed:
Review fixes

evenh committed:
Update AGENTS.md to reflect SKIPJob v1beta1

andtag commented on internal/controllers/common/util.go:
```diff
@@ -8,8 +8,8 @@ import (
-	skiperatorv1alpha1 "github.com/kartverket/skiperator/api/v1alpha1"
-	"github.com/kartverket/skiperator/api/v1alpha1/podtypes"
+	"github.com/kartverket/skiperator/api/common"
+	"github.com/kartverket/skiperator/api/common/podtypes"
```
Trenger vi linje 12?

berg-thom commented on cmd/skiperator/main.go:
```diff
@@ -101,27 +111,67 @@ func main() {
+
+		// Initialize webhook certificate watcher if webhook cert directory is provided
+		if len(*webhookCertDir) > 0 {
+			certPath := filepath.Join(*webhookCertDir, *webhookCertName)
```
Er det kanskje fornuftig å sjekke om pathen som angis faktisk er en valid path  ala	 `fs.ValidPath(certpath)`, eller er det for paranoid?

Morisman96 commented on CONTRIBUTING.md:
Worked for me in IntelliJ, maybe worth documenting?
```suggestion
>NOTE: If you want to run Skiperator in debug mode, use `make local-webhook` instead of `make run-local`. Then, add the output target directory (e.g., `/tmp/skiperator-webhook-certs.XXXXXX`) as argument: `--webhook-cert-dir=/tmp/skiperator-webhook-certs.XXXXXX --webhook-host=0.0.0.0`
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```suggestion
func (r *ReconcilerBase) updateStatus(ctx context.Context, skipObj common.SKIPObject) {
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```suggestion
func (r *ReconcilerBase) SetSyncedState(ctx context.Context, skipObj common.SKIPObject, message string) {
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```suggestion
func (r *ReconcilerBase) SetProgressingState(ctx context.Context, skipObj common.SKIPObject, message string) {
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```suggestion
func (r *ReconcilerBase) SetErrorState(ctx context.Context, skipObj common.SKIPObject, err error, message string, reason string) {
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```suggestion
		latestObj := skipObj.DeepCopyObject().(common.SKIPObject)
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```suggestion
func (r *ReconcilerBase) UpdateAccessPolicy(ctx context.Context, obj common.SKIPObject) {
```

Morisman96 commented on internal/controllers/common/reconciler.go:
```diff
@@ -4,8 +4,8 @@ import (
 	"fmt"
 
+	"github.com/kartverket/skiperator/api/common/podtypes"
 	"github.com/kartverket/skiperator/api/v1alpha1"
```
```suggestion
	"github.com/kartverket/skiperator/api/common"
```

Morisman96 reviewed (commented):
Bruke common isteden for v1alpha1 for mer lesbarhet, skal ikke ha noe å si kode messig siden  `type SKIPObject = commontypes.SKIPObject` i v1alpha1, så det er bare litt småpirk
